src/blockdeque.jl

```diff
@@ -0,0 +1,152 @@
+"""
+    append!(dest::BlockVector, sources...; alias = false)
+
+Append items from `sources` to the last block of `dest`.  If `alias = true`,
+append the blocks in `sources` as-is if they are compatible with the internal
+block array type of `dest`.  Importantly, this means that mutating `sources`
+afterwards alters the items in `dest` and it may even break the invariance
+of `dest` if the length of `sources` are changed.  The elements may be copied
+even if `alias = true` if the corresponding implementation does not exist.
+
+The blocks in `dest` must not alias with `sources` or components of them.
+For example, the result of `append!(x, x)` is undefined.
+"""
+Base.append!(dest::BlockVector, sources...; alias::Bool = false) =
+    foldl((d, s) -> append!(d, s; alias = alias), sources; init = dest)
+
+function Base.append!(dest::BlockVector, src; alias::Bool = false)
+    if _blocktype(dest) === _blocktype(src) && alias
+        return append_nocopy!(dest, src)
+    else
+        return append_copy!(dest, src)
+    end
+end
+
+_blocktype(::Any) = nothing
+_blocktype(::T) where {T<:AbstractArray} = T
+_blocktype(::BlockArray{<:Any,<:Any,<:AbstractArray{T}}) where {T<:AbstractArray} = T
+_blocktype(::PseudoBlockArray{<:Any,<:Any,T}) where {T<:AbstractArray} = T
+
+function append_nocopy!(dest::BlockVector{<:Any,T}, src::BlockVector{<:Any,T}) where {T}
+    isempty(src) && return dest
+    append!(dest.blocks, src.blocks)
+    offset = last(dest.axes[1]) + 1 - src.axes[1].first
+    append!(dest.axes[1].lasts, (n + offset for n in src.axes[1].lasts))
+    return dest
+end
+
+append_nocopy!(
+    dest::BlockVector{<:Any,<:AbstractArray{T}},
+    src::PseudoBlockVector{<:Any,T},
+) where {T} = append_nocopy!(dest, src.blocks)
+
+function append_nocopy!(dest::BlockVector{<:Any,<:AbstractArray{T}}, src::T) where {T}
+    isempty(src) && return dest
+    push!(dest.blocks, src)
+    push!(dest.axes[1].lasts, last(dest.axes[1]) + length(src))
+    return dest
+end
+
+append_copy!(dest::BlockVector, src) = _append_copy!(dest, Base.IteratorSize(src), src)
+
+function _append_copy!(dest::BlockVector, ::Union{Base.HasShape,Base.HasLength}, src)
+    block = dest.blocks[end]
+    li = lastindex(block)
+    resize!(block, length(block) + length(src))
+    # Equivalent to `i = li; for x in src; ...; end` but (maybe) faster:
+    foldl(src, init = li) do i, x
+        Base.@_inline_meta
+        i += 1
+        @inbounds block[i] = x
+        return i
+    end
```

Comment:
It's faster when `src` is a block vector.

With this PR:

```
julia> @btime append!(x, $(mortar([rand(3) for _ in 1:100]))) setup=(x = mortar([[0.0]]));
  912.130 ns (5 allocations: 9.29 KiB)
```

Without the specific `foldl` implementation added in this PR:

```
julia> @btime append!(x, $(mortar([rand(3) for _ in 1:100]))) setup=(x = mortar([[0.0]]));
  11.031 μs (7 allocations: 2.58 KiB)
```